File: www/cgi/session.pl
#!/usr/bin/perl
use strict;
use warnings;
use File::Path qw(make_path);
use File::Spec;

# 1. Configuração do diretório de sessões
my $session_dir = "/tmp/webserv_sessions";
unless (-d $session_dir) {
    make_path($session_dir);
}

# 2. Parse do Cookie
my $cookie_header = $ENV{'HTTP_COOKIE'} || "";
my %cookies;
foreach (split /;/, $cookie_header) {
    my ($k, $v) = split /=/, $_, 2;
    if (defined $k && defined $v) {
        $k =~ s/^\s+|\s+$//g;
        $cookies{$k} = $v;
    }
}

my $session_id = $cookies{'session_id'};
my $visits = 0;

# 3. Validação/Criação da Sessão
if ($session_id && $session_id =~ /^[a-zA-Z0-9\-]+$/) {
    my $session_file = File::Spec->catfile($session_dir, "session_$session_id");
    if (-e $session_file) {
        if (open(my $fh, '<', $session_file)) {
            $visits = <$fh>;
            chomp($visits);
            close($fh);
            $visits = int($visits || 0);
        }
    } else {
        $session_id = generate_uuid();
        $visits = 0;
    }
} else {
    $session_id = generate_uuid();
    $visits = 0;
}

$visits++;

# 4. Salvar sessão
my $session_file = File::Spec->catfile($session_dir, "session_$session_id");
if (open(my $fh, '>', $session_file)) {
    print $fh $visits;
    close($fh);
}

# 5. Output HTTP Headers
print "Content-Type: text/html\r\n";
print "Set-Cookie: session_id=$session_id; Path=/; HttpOnly\r\n";
print "\r\n";

# 6. Output Body
print <<"HTML";
<html>
<head><title>Webserv Session Test (Perl)</title></head>
<body>
    <h1>Session ID: $session_id</h1>
    <h2>Visit Count: $visits</h2>
    <p>Script executado via CGI Perl!</p>
    <p>Refresh para incrementar o contador.</p>
    <p><a href='/'>Voltar para Home</a></p>
</body>
</html>
HTML

# Função simples para gerar um ID aleatório
sub generate_uuid {
    my @chars = ('a'..'z', 'A'..'Z', 0..9);
    my $uuid = "";
    foreach (1..16) { $uuid .= $chars[rand @chars]; }
    return $uuid;
}

File: www/cgi/session.py
#!/usr/bin/env python3
import os
import sys
import json
import uuid
import hashlib

SESSION_DIR = "/tmp/webserv_sessions"


def session_file(session_id):
    return os.path.join(SESSION_DIR, hashlib.md5(session_id.encode()).hexdigest() + ".json")


def parse_cookies(cookie_header):
    cookies = {}
    for part in cookie_header.split(";"):
        part = part.strip()
        if "=" in part:
            key, value = part.split("=", 1)
            cookies[key.strip()] = value.strip()
    return cookies


def main():
    if not os.path.isdir(SESSION_DIR):
        os.makedirs(SESSION_DIR)

    cookies = parse_cookies(os.environ.get("HTTP_COOKIE", ""))
    session_id = cookies.get("session_id", "")
    new_session = False
    visits = 0

    if session_id:
        try:
            with open(session_file(session_id), "r") as handle:
                data = json.load(handle)
                visits = int(data.get("visits", 0))
        except Exception:
            session_id = str(uuid.uuid4())
            new_session = True
            visits = 0
    else:
        session_id = str(uuid.uuid4())
        new_session = True

    visits += 1
    with open(session_file(session_id), "w") as handle:
        json.dump({"visits": visits}, handle)

    body = {
        "session_id": session_id,
        "visits": visits,
        "html": "Session ID: " + session_id + "\nVisit Count: " + str(visits)
    }
    encoded = json.dumps(body)

    sys.stdout.write("Content-Type: application/json\r\n")
    if new_session:
        sys.stdout.write("Set-Cookie: session_id=" + session_id + "; Path=/; HttpOnly\r\n")
    sys.stdout.write("Content-Length: " + str(len(encoded)) + "\r\n")
    sys.stdout.write("\r\n")
    sys.stdout.write(encoded)


if __name__ == "__main__":
    main()
